### PR TITLE
refactor: migrate from argon2 -> @node-rs/argon2

### PR DESCRIPTION
**Pull request by jsjoeio**
This migrates our hashing password implementation from using `argon2` to `@node-rs/argon2`.

Fixes https://github.com/coder/code-server/issues/4422

## How to Test

1. Wait for [CI](https://github.com/coder/code-server/actions/runs/1690093340) to finish
2. Download `release-packages` from Artifacts 
3. Unzip and then run `cd` into `release-packages`
4. Run `tar -xzf code-server*` based on your arch
5. Generate an argon2 password or use this one: "$argon2i$v=19$m=4096,t=3,p=1$cecR+o8tshB6mKzPBUG7fw$w2XFt+Ezn+CjRGTwptsjH3yj72htGLhMc3WdPs7wIOk" (equals "password")
6. edit your config.yaml and add `hashed-password: "$argon2i$v=19$m=4096,t=3,p=1$cecR+o8tshB6mKzPBUG7fw$w2XFt+Ezn+CjRGTwptsjH3yj72htGLhMc3WdPs7wIOk"`
7. Run `./code-server*/bin/code-server` 
8. Login with password: "password"
9. Everything should work

### Tested on

- [x] macos @jsjoeio 
- [x] linux @jsjoeio 
- [x] Termux @jsjoeio 
- [ ] Raspberry Pi @im-coder-lg 

## Notes

At first, I ran into some issues with tests because I was overriding `process.platform` and that caused issues with any test files that required `@node-rs/argon2`. @code-asher had a good idea to fix this: refactor `getEnvPaths()` to pass in `process.platform`. That fixed it.

### Resources

- https://crates.io/crates/argon2
- https://github.com/napi-rs/node-rs/tree/main/packages/argon2
- https://github.com/coder/code-server/issues/4422#issuecomment-1005335764

### CODE REVIEW

File: src/node/util.ts
Comment (jsjoeio) [inline src/node/util.ts:179]: 🧹 I realized it didn't make sense to throw when checking if the hash and password were a match so I refactored to log the error and return `false`.

File: ci/build/build-standalone-release.sh
Comment (jsjoeio) [inline ci/build/build-standalone-release.sh:5]: ❓ This is legacy code. I'm not sure if we still need to tell npm to build from source 🤔

cc @yisibl any ideas?

File: ci/build/npm-postinstall.sh
Comment (jsjoeio) [inline ci/build/npm-postinstall.sh:22]: Same as above cc @yisibl

File: src/node/util.ts
Comment (jsjoeio) [inline src/node/util.ts:61]: ✍🏼 This change was suggested by @code-asher 🔥

By passing in `process.platform`, we can override it in our tests without having to do hacky stuff like modify `process.platform` (which I was doing before).

File: test/unit/node/util.test.ts
Comment (code-asher) [inline test/unit/node/util.test.ts:143]: Looks like this should be updated as well (false instead of throw).

File: src/node/util.ts
Comment (jsjoeio) [reply src/node/util.ts:2]: Oh great point! Yes, I don't have a strong preference, I only went with this approach at first since I was copying their examples.

File: test/unit/node/util.test.ts
Comment (jsjoeio) [reply test/unit/node/util.test.ts:143]: Nice catch!

File: src/node/util.ts
Comment (yisibl) [reply src/node/util.ts:179]: Nice change.

File: ci/build/build-standalone-release.sh
Comment (yisibl) [reply ci/build/build-standalone-release.sh:5]: Theoretically not needed, node-rs/argon2 already compiles the binary (.node) file for CentOS.

**Comment by im-coder-lg** [comment]
@yisibl any ideas why the artifact failed?

**Comment by yisibl** [comment]
> @yisibl any ideas why the artifact failed?

 What is the actual glibc version in Docker?

**Comment by Brooooooklyn** [comment]
> @yisibl any ideas why the artifact failed?

`@node-rs/argon2` doesn't support `GLIBC_2.17` for now. I'm implementing the support in https://github.com/napi-rs/napi-rs/pull/1025

And I'll draft a patch release after `NAPI-RS` toolchain supported the lower `GLIBC` on the Linux platform.

**Comment by yisibl** [comment]
Yeah, I see `GLIBC_2.17`,  let's wait for the next version.

**Comment by im-coder-lg** [comment]
@yisibl is the new version out? I see a new release dating of 4h as of typing this...

**Comment by im-coder-lg** [comment]
Actually, idk if I could test on Termux, since whenever I tried it it didn't work. But I will try 3.9.3 and try learning how to do that.

**Comment by im-coder-lg** [comment]
Uh oh... My Termux isn't working. From the beginning, I haven't able to get the mirrors updated, there isn't a sources.list file in my /etc dir. Idk what happened to my phone, maybe it's because I have an older phone.

**Comment by im-coder-lg** [comment]
Wait we are supposed to use `@napi-rs/argon2` right? Whys it node-rs?

**Comment by Brooooooklyn** [comment]
Upgrade to `@node-rs/argon2@1.0.4` to solve the `GLIBC` problem with `centos:7`

File: ci/build/npm-postinstall.sh
Comment (jsjoeio) [reply ci/build/npm-postinstall.sh:22]: can remove

**Comment by jsjoeio** [comment]
@Brooooooklyn will do - thank you for the quick patch!

@yisibl - thank you for the review and feedback!

@im-coder-lg no worries, I will test on Termux

**Comment by jsjoeio** [comment]
This should fix the audit/code scanning CI failing (I'll rebase once merged): https://github.com/coder/code-server/pull/4742

**Comment by jsjoeio** [comment]
works on macOS!

**Comment by im-coder-lg** [comment]
@jsjoeio the total size of the build artifacts are nearing 2 GB. How are you all managing it?
Anyway, I'll test it. I am booting my PC now to download the artifact and then extract it. Once that's done, I'll use WinSCP to transfer the file and test on my RasPi.

**Comment by im-coder-lg** [comment]
Testing this _now_!

**Comment by jsjoeio** [comment]
> the total size of the build artifacts are nearing 2 GB. How are you all managing it?

😅 I wish I had a good answer for that...

**Comment by jsjoeio** [comment]
@code-asher brought up a good point. our e2e tests run on Linux so those are passing which means this is good on Linux.

**Comment by jsjoeio** [comment]
@im-coder-lg any updates?

**Comment by jsjoeio** [comment]
## Termux

Previously, installing the project would fail on Termux due to argon2 issues. From what I can tell, it seems as though those problems are resolved with this PR!

Screenshot_20220114-121950

I couldn't successfully downgrade Node to v14 on Termux so code-server wouldn't actually run. So assuming someone can downgrade, it _should_ work.

**Review by code-asher — APPROVED** [review]

**Comment by im-coder-lg** [comment]
I'll just test again. Since I didn't install Node 14 before running, I got into errors. I will try again and hopefully succeed.

**Comment by jsjoeio** [comment]
> I'll just test again. Since I didn't install Node 14 before running, I got into errors. I will try again and hopefully succeed.

No worries at all. What I did was upload the `npm-package` to Google Drive, then download directly on my device, unzip/untar and then `yarn` in that (`yarn install --production`) and was able to get it to work.

**Comment by jsjoeio** [comment]
Here's a link to the npm-package.zip if you want to download: https://drive.google.com/file/d/1VyB0i4GMf9yAo8n_DU-YkfPGRKnzb9Nf/view?usp=sharing

**Comment by im-coder-lg** [comment]
Heads up: I'll be testing two architectures, the first one is AARCH64 aka. ARM64 and then ARMV7L which is integrated to Raspberry Pi OS Bullseye.

**Comment by im-coder-lg** [comment]
LOLZ, all I needed to test was the `package.tar.gz` tarball? I couldn't unzip the ZIP file, maybe it's due to that.

**Comment by im-coder-lg** [comment]
[This](https://github.com/coder/code-server/pull/4733#issuecomment-1013453136) absolutely works. I am testing this once again with the hashed password and maybe again on armv7l.

**Comment by im-coder-lg** [comment]
I can say this works successfully, although 4.0.1 is still in development, some simple errors pop up.
Logged in, opened folders, tested the typing, everything works.
Maybe now I'll try a simple webserver. Docusaurus v2?

**Review by im-coder-lg — APPROVED** [review]
We can merge! LGTM!

File: package.json
Comment (jsjoeio) [reply package.json]: Yes! I checked and I've updated it to 1.0.5!

**Comment by jsjoeio** [comment]
@im-coder-lg thanks so much for helping test!!! I will merge this once CI passes

**Pull request merged**
